Cython/Compiler/PyrexTypes.py

```diff
@@ -4398,6 +4404,107 @@ def error_condition(self, result_code):
         return "dummy"
 
 
+class PythonTypeConstructor(PyObjectType):
+    """Used to help Cython interpret indexed types from the typing module (or similar)
+    """
+    is_python_type_constructor = True
+
+    def __init__(self, name, base_type=None):
+        self.name = name
+        self.base_type = base_type
+
+    def specialize_here(self, pos, env, template_values=None):
+        if self.base_type:
+            # for a lot of the typing classes it doesn't really matter what the template it
+            # (i.e. typing.Dict[int] is really just a dict)
+            return self.base_type
+        return self
+
+    def __repr__(self):
+        if self.base_type:
+            return "%s[%r]" % (self.name, self.base_type)
+        else:
+            return self.name
+
+    def is_template_type(self):
+        return True
+
+
+class PythonTupleTypeConstructor(PythonTypeConstructor):
+    def specialize_here(self, pos, env, template_values=None):
+        if (template_values and None not in template_values and
+                not any(v.is_pyobject for v in template_values)):
+            entry = env.declare_tuple_type(pos, template_values)
+            if entry:
+                return entry.type
+        return super(PythonTupleTypeConstructor, self).specialize_here(pos, env, template_values)
+
+
+class SpecialPythonTypeConstructor(PythonTypeConstructor):
+    """
+    For things like ClassVar, Optional, etc, which have extra features on top of being
+    a "templated" type.
+    """
+    is_special_python_type_constructor = True
+
+    @property
+    def is_python_type_constructor(self):
+        return not bool(self.template_type)
+
+    def __init__(self, name, template_type=None):
+        super(SpecialPythonTypeConstructor, self).__init__(name, None)
+        if (name == "typing.ClassVar" and template_type
+                and not template_type.is_pyobject):
+            # because classvars end up essentially used as globals they have
+            # to be PyObjects. Try to find the nearest suitable type (although
+            # practically I doubt this matters).
+            py_type_name = template_type.py_type_name()
+            if py_type_name:
+                from .Builtin import builtin_scope
+                template_type = (builtin_scope.lookup_type(py_type_name)
+                                        or py_object_type)
+            else:
+                template_type = py_object_types
+        self.template_type = template_type
+
+    def __getattribute__(self, attr):
```

Comment:
I think I've successfully got rid of it. Instead of having `specialize` return a `SpecialPythonTypeConstructor` that uses `__getattribute__` to try to look like the underlying type, I've had `specialize` return a copy of the underlying type with a few identifying attributes set.

Returning a copy isn't great because isn't a lot of types are tested by `is`. But the old mechanism didn't work there either, and it's somewhat mitigated by using `resolve` to retrieve the copied type.